tachyon/zk/air/constraint_system/constraint_system.h

```diff
@@ -0,0 +1,94 @@
+#ifndef TACHYON_ZK_AIR_CONSTRAINT_SYSTEM_CONSTRAINT_SYSTEM_H_
+#define TACHYON_ZK_AIR_CONSTRAINT_SYSTEM_CONSTRAINT_SYSTEM_H_
+
+#include <stddef.h>
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "tachyon/math/matrix/matrix_types.h"
+#include "tachyon/zk/air/constraint_system/variable.h"
+#include "tachyon/zk/air/expressions/air_evaluator.h"
+
+namespace tachyon::zk::air {
+
+template <typename F>
+class ConstraintSystem {
+ public:
+  ConstraintSystem() = default;
+  ConstraintSystem(size_t num_public_values, size_t main_width,
+                   size_t preprocessed_width)
+      : num_public_values_(num_public_values),
+        main_width_(main_width),
+        preprocessed_width_(preprocessed_width) {}
+
+  void EnforceFirstRowConstraint(std::unique_ptr<Expression<F>> constraint) {
+    EnforceConstraint(ExpressionFactory<F>::FirstRow(std::move(constraint)));
+  }
+
+  void EnforceLastRowConstraint(std::unique_ptr<Expression<F>> constraint) {
+    EnforceConstraint(ExpressionFactory<F>::LastRow(std::move(constraint)));
+  }
+
+  void EnforceTransitionConstraint(std::unique_ptr<Expression<F>> constraint) {
+    EnforceConstraint(ExpressionFactory<F>::Transition(std::move(constraint)));
+  }
+
+  void EnforceConstraint(std::unique_ptr<Expression<F>> constraint) {
+    constraints_.push_back(std::move(constraint));
+  }
+
+  bool IsSatisfied(
+      AirEvaluator<F>& evaluator, const std::vector<F>& public_values,
+      const math::RowMajorMatrix<F>& main,
+      const math::RowMajorMatrix<F>* preprocessed = nullptr) const {
+    CHECK_EQ(public_values.size(), num_public_values_);
+    CHECK_EQ(main.cols(), main_width_);
+    if (preprocessed) {
+      CHECK_EQ(preprocessed->cols(), preprocessed_width_);
+      CHECK_EQ(preprocessed->rows(), main.rows());
+    } else {
+      CHECK_EQ(preprocessed_width_, 0);
+    }
+
+    evaluator.set_public_values(public_values);
+    evaluator.set_num_rows(main.rows());
+
+    for (Eigen::Index i = 0; i < main.rows() - 1; ++i) {
+      const math::RowMajorMatrix<F>& main_window = main.middleRows(i, 2);
```

Comment:
If you take this as `Eigen::Matrix`, this will cause a copy. Please modify the type to `Eigen::Block<>`.
If you take this as `Eigen::Matrix`, this will cause a copy. Please modify the type to `Eigen::Block<>`.

```suggestion
      Eigen::Block<math::RowMajorMatrix<F>> main_window = main.middleRows(i, 2);
```